### ISSUE TITLE
Streamlit Dataset Dumping Issue

### ISSUE THREAD

**Issue by Xu-Justin** [danilhendrasr/semi-automated-labeling#40]
Exporting the same dataset from CVAT to dump **twice**, cause the dataset unable to load any sample (`Num samples: 0`).

![image](https://user-images.githubusercontent.com/56458008/176608511-eb40837c-470a-49a1-b686-e2486ebd2096.png)

**Comment by ruhyadi** [comment]
Ini dibagian mana?

**Comment by Xu-Justin** [comment]
![image](https://user-images.githubusercontent.com/56458008/176609828-fb9afa17-83de-42f4-8954-ad0b96fb5faa.png)

Di halaman ini ketika tombol `Preview in Fiftyone` ditekan dua kali.

**Comment by Xu-Justin** [comment]
![image](https://user-images.githubusercontent.com/56458008/176610223-c2b9b1bc-5948-4e26-a325-e5c6c490c78e.png)

Log dari streamlit.

**Comment by ruhyadi** [comment]
Semua datasetnya harus dihapus dulu. Good practice untuk satu dataset per sesi FiftyOne.

```
list_datasets = fo.list_datasets()
try:
    [fo.delete_dataset(name) for name in list_datasets]
except:
    print("[INFO] No Existing Dataset")
```

**Comment by Xu-Justin** [comment]
Ga ada perubahan. Sepertinya pengaruh di dumping directorynya, bukan di storage fiftyone.

**Comment by ruhyadi** [comment]
Maybe, tapi sekarang aku coba gak ada masalah (belum di env docker). Di Fiftyone juga bisa ke load juga yg keduanya.

![image](https://user-images.githubusercontent.com/39303602/176614075-8451e09b-0107-4e70-a6ed-a2df4fdd55e2.png)

**Comment by Xu-Justin** [comment]
> Ga ada perubahan. Sepertinya pengaruh di dumping directorynya, bukan di storage fiftyone.

Further inspection shows that the problem is at `fo.Dataset.from_dir`, not the dumping directory.

```py
>>> fo.list_datasets()
['2022.07.04.05.06.03', '39']
```
```py
>>> fo.delete_dataset(name)
>>> fo.list_datasets()
['2022.07.04.05.06.03']
>>> fo.Dataset.from_dir(
    dataset_dir = dataset_dir,
    dataset_type = fo.types.COCODetectionDataset,
    name = name
)
Name:        39
Media type:  image
Num samples: 50
Persistent:  False
Tags:        []
Sample fields:
    id:           fiftyone.core.fields.ObjectIdField
    filepath:     fiftyone.core.fields.StringField
    tags:         fiftyone.core.fields.ListField(fiftyone.core.fields.StringField)
    metadata:     fiftyone.core.fields.EmbeddedDocumentField(fiftyone.core.metadata.ImageMetadata)
    ground_truth: fiftyone.core.fields.EmbeddedDocumentField(fiftyone.core.labels.Detections)

>>> fo.list_datasets()
['2022.07.04.05.06.03', '2022.07.04.05.06.15', '39']
```

Notice that `fo.Dataset.from_dir` create two dataset `2022.07.04.05.06.15` and `39`.
(EDIT: Solved on #46 )

And something interesting shows here.

```py
>>> fo.load_dataset('39')
Name:        39
Media type:  image
Num samples: 0
Persistent:  False
Tags:        []
Sample fields:
    id:           fiftyone.core.fields.ObjectIdField
    filepath:     fiftyone.core.fields.StringField
    tags:         fiftyone.core.fields.ListField(fiftyone.core.fields.StringField)
    metadata:     fiftyone.core.fields.EmbeddedDocumentField(fiftyone.core.metadata.ImageMetadata)
    ground_truth: fiftyone.core.fields.EmbeddedDocumentField(fiftyone.core.labels.Detections)
```
```py
>>> fo.load_dataset('2022.07.04.05.06.15')
Name:        2022.07.04.05.06.15
Media type:  image
Num samples: 50
Persistent:  False
Tags:        []
Sample fields:
    id:           fiftyone.core.fields.ObjectIdField
    filepath:     fiftyone.core.fields.StringField
    tags:         fiftyone.core.fields.ListField(fiftyone.core.fields.StringField)
    metadata:     fiftyone.core.fields.EmbeddedDocumentField(fiftyone.core.metadata.ImageMetadata)
    ground_truth: fiftyone.core.fields.EmbeddedDocumentField(fiftyone.core.labels.Detections)
```

**Comment by ruhyadi** [comment]
Betul, coba open dataset `'2022.07.04.05.06.15'` di FiftyOne ada isinya gak? (Bisa keliatan isinya gak). 
Klo diaku isinya gak bisa diliat, kebaca si berapa jumlah GT nya tapi isinya gak bisa diliat.

![image](https://user-images.githubusercontent.com/39303602/177091980-4c163b5f-be91-4ead-afbd-48f8769313f4.png)

Dan karena fungsi yg Post Annotations (Label Evalutor) yg tanpa Embedding itu pake Fungsi `fiftyone51.preview_fiftyone()`, dan return dari fungsi yg dipake cuman `patch` nya ajah. Jadi image yg ilang didataset gak ngaruh sama keseluruhan fungsinya (masih bisa bekerja untuk ubah label). 

Btw, fungsi `fiftyone51.preview_fiftyone()` itu nerapin fungsi hapus semua dataset, jadi yg dataset fiftyone macem `'2022.07.04.05.06.15'` gak ada, karena udah dihapus duluan. 

Kayaknya ini emng bugs dari `fo.Dataset.from_dir()`

**Comment by Xu-Justin** [comment]
I tried to run `fo.Dataset.from_dir()` as normal python script (without Streamlit), and everything is fine.

~~Hypothesis: Streamlit multithreading service (needed for auto-run/auto-rerun feature) causes Fiftyone database to misbehave (concurrency error).~~

**Comment by Xu-Justin** [comment]
Problem found! Main problem on inconsistent data between Fiftyone database and CVAT task dumping.

I'll update the solution soon.

**Comment by Xu-Justin** [comment]
Solved.

# Solution
Adding `fo.list_datasets()` before `fo.load_dataset()` solved the problems.
```py
def load_dataset(name):
    fo.list_datasets()
    return fo.load_dataset(name)
```

EDIT: Not working all the time.

# Explanation
I have no idea.

![image](https://user-images.githubusercontent.com/56458008/177142377-40c600c6-c60a-40e7-8c36-9d76365a6b9f.png)

**Comment by Xu-Justin** [comment]
Problem still occures.

**Comment by ruhyadi** [comment]
Mungkin gak klo kita biarin ajah bugs di FiftyOne, dan ngubah mekanisme plotly nya?. Coba cek ini: [https://github.com/null-jones/streamlit-plotly-events](https://github.com/null-jones/streamlit-plotly-events )

**Comment by Xu-Justin** [comment]
Solved by #55
